packages/queued-request-controller/src/QueuedRequestMiddleware.ts

```diff
@@ -0,0 +1,131 @@
+    'wallet_watchAsset',
+    'wallet_switchEthereumChain',
+    'eth_signTypedData_v4',
+  ];
+
+  return !confirmationMethods.includes(method);
+};
+
+export const createQueuedRequestMiddleware = (
+  messenger: ControllerMessenger<
+    | QueuedRequestControllerEnqueueRequestAction
+    | NetworkControllerGetStateAction
+    | NetworkControllerSetActiveNetworkAction
+    | NetworkControllerSetProviderTypeAction
+    | NetworkControllerGetNetworkClientByIdAction
+    | NetworkControllerFindNetworkClientIdByChainIdAction
+    | SelectedNetworkControllerSetNetworkClientIdForDomainAction
+    | AddApprovalRequest,
+    never
+  >,
+  useRequestQueue: () => boolean,
+): JsonRpcMiddleware<any, any> => {
+  return createAsyncMiddleware(async (req: any, res: any, next) => {
+    if (!useRequestQueue() || isConfirmationMethod(req.method)) {
+      next();
+      return;
+    }
+
+    const networkClientIdForRequest = req.networkClientId as NetworkClientId;
+
+    await messenger.call(
+      QueuedRequestControllerActionTypes.enqueueRequest,
+      async () => {
+        if (req.method === 'wallet_switchEthereumChain') {
+          // eslint-disable-next-line n/callback-return
+          return next();
+        }
+
+        const builtIn =
+          BUILT_IN_NETWORKS[
+            networkClientIdForRequest as keyof typeof BUILT_IN_NETWORKS
+          ];
+        const isBuiltIn = builtIn !== undefined && builtIn.chainId;
+        const networkConfigurationForRequest = isBuiltIn
+          ? builtIn
+          : messenger.call(
+              'NetworkController:getNetworkClientById',
+              networkClientIdForRequest,
+            ).configuration;
+
+        const currentProviderConfig = messenger.call(
+          'NetworkController:getState',
+        ).providerConfig;
+
+        const currentChainId = currentProviderConfig.chainId;
+
+        if (currentChainId === networkConfigurationForRequest.chainId) {
+          // eslint-disable-next-line n/callback-return
+          return next();
+        }
+
+        // if is switch eth chain call
+        // clear request queue when the switch ethereum chain call completes (success or fail)
+        // This is because a dapp-requested switch ethereum chain invalidates any requests they've made after this switch, since we dont know if they were expecting the chain after the switch or before.
+        // with the queue batching approach, this would mean clearing any batch for that origin (batches being per-origin.)
+        const requestData = {
+          toNetworkConfiguration: networkConfigurationForRequest,
+          fromNetworkConfiguration: currentProviderConfig,
+        };
+
+        try {
+          const approvedRequestData = (await messenger.call(
+            'ApprovalController:addRequest',
+            {
+              origin,
+              type: ApprovalType.SwitchEthereumChain,
+              requestData,
+            },
+            true,
+          )) as any;
```

Comment:
is this cast necessary?

Reply:
the resulting type is 'Unknown' - casting it to any isnt great though...

Reply:
could do something like this instead, which still is casting, but is more specific:
```
          const approvedRequestData = (await messenger.call(
            'ApprovalController:addRequest',
            {
              origin,
              type: ApprovalType.SwitchEthereumChain,
              requestData,
            },
            true,
          ))

          if (isBuiltIn) {
            await messenger.call(
              'NetworkController:setProviderType',
              (approvedRequestData as { type: InfuraNetworkType }).type,
            );
          } else {
            await messenger.call(
              'NetworkController:setActiveNetwork',
              (approvedRequestData as { id: NetworkClientId}).id,
            );
          }
```